### PR TITLE
Allow initializing client.Batch outside of txn retryable

### LINKED ISSUE #3067: initializing client.Batch outside of txn retryable is dangerous
This here will go horribly wrong (endless retry loop) unless the retryable succeeds on its first attempt.

```go
b := &client.Batch{}
b.Put(key, "test")

s.db.Txn(func(txn *client.Txn) error {
        return txn.CommitInBatch(b)
})
```

Instead, the batch needs to be initialized inside of the retryable (or the error stored within it cleared manually). I ran into this in WIP and took me a bit to track it down. We should reset errors in an appropriate location so that the above isn't a death trap.

### PR DESCRIPTION
**Pull request by yaojingguo**
Fixes #3067

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3203)

### CODE REVIEW

**Comment by petermattis** [comment]
@yaojingguo Can you provide a bit more justification for this change? The main benefit seems to be avoiding re-initializing a batch upon txn retry. But that should be rare. Do you have a benchmark that shows this to be a worthwhile benefit. I'm mildly concerned that you haven't fully reset the batch state (e.g. cleared errs) and doing so will be a bit tricky because `Result.Err` is populated both during batch creation and batch execution.

**Comment by tschottdorf** [comment]
I think the only concern is #3067.

**Comment by petermattis** [comment]
Ah, I failed to look at the issue. My bad.

#3067 doesn't require that we allow initialization of a batch outside of a transaction, only that we detect the error. There are several possibilities here. We could add a flag to `client.Batch` to detect when it has been run and refuse to allow it to be run again. Or we could add a `txn.NewBatch()` method and only allow the batch to be run on the `txn` that created it.

**Comment by tschottdorf** [comment]
The first option doesn't seem like the right fix (it wouldn't have caught
the error when the code was written or exercised, only when I managed to
change something unrelated that actually got that Txn retrying). The second
would work well.

**Comment by yaojingguo** [comment]
@petermattis, a new commit to implement the second option. The build fails because I have not updated the code locations which invoke `CommitInBatchWithResponse` directly or indirectly.  If you think that this commit is OK, I will proceed to update these code locations .

**Comment by yaojingguo** [comment]
@petermattis, any suggestions?

**Comment by petermattis** [comment]
This approach looks good to me.

**Comment by yaojingguo** [comment]
A new commit after updating the aforementioned code locations. Since there are some acceptance test failures in master branch,  the commit does not pass `ci/circleci` check. But I have successfully run all the unit test cases in related packages.

**Comment by tschottdorf** [comment]
A rebase should fix those acceptance failures.

**Comment by yaojingguo** [comment]
> A rebase should fix those acceptance failures.

I have rebased. The first attempt for master branch by `ci/circleci` have failed. See https://circleci.com/gh/cockroachdb/cockroach/9669.

**Comment by tschottdorf** [comment]
yes, that's another flaky test. See #3219, should hopefully not be an issue any more soon. In the meantime, try another restart.

**Comment by petermattis** [comment]
LGTM

**Comment by yaojingguo** [comment]
A new commit after a git rebase and updating a new code location in `acceptance/single_key_test.go` .It has passed `ci/circleci` check.

File: client/batch.go [hunk 1]
```diff
@@ -33,6 +33,9 @@ type Batch struct {
 	// The DB the batch is associated with. This field may be nil if the batch
 	// was not created via DB.NewBatch or Txn.NewBatch.
 	DB *DB
+	// The Txn the batch is associated with. This field may be nil if the batch
+	// was not created via Txn.NewBatch.
+	Txn *Txn
```
Comment (tamird) [inline client/batch.go]: this doesn't need to be exported

**Comment by yaojingguo** [comment]
@tamird. A new commit after changing `Txn` to `txn` , a git rebase and updating some new code locations which use `CommitInBatchWithResponse`.

**Comment by tamird** [comment]
LGTM

**Pull request merged**